// Data.Tuple.Nested/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Tuple = require("Data.Tuple");
var $div$bslash = function (a) {
    return function (b) {
        return new Data_Tuple.Tuple(a, b);
    };
};
var con9 = function (f) {
    return function (_213) {
        return f(_213.value0)(_213.value1.value0)(_213.value1.value1.value0)(_213.value1.value1.value1.value0)(_213.value1.value1.value1.value1.value0)(_213.value1.value1.value1.value1.value1.value0)(_213.value1.value1.value1.value1.value1.value1.value0)(_213.value1.value1.value1.value1.value1.value1.value1.value0)(_213.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con8 = function (f) {
    return function (_212) {
        return f(_212.value0)(_212.value1.value0)(_212.value1.value1.value0)(_212.value1.value1.value1.value0)(_212.value1.value1.value1.value1.value0)(_212.value1.value1.value1.value1.value1.value0)(_212.value1.value1.value1.value1.value1.value1.value0)(_212.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con7 = function (f) {
    return function (_211) {
        return f(_211.value0)(_211.value1.value0)(_211.value1.value1.value0)(_211.value1.value1.value1.value0)(_211.value1.value1.value1.value1.value0)(_211.value1.value1.value1.value1.value1.value0)(_211.value1.value1.value1.value1.value1.value1);
    };
};
var con6 = function (f) {
    return function (_210) {
        return f(_210.value0)(_210.value1.value0)(_210.value1.value1.value0)(_210.value1.value1.value1.value0)(_210.value1.value1.value1.value1.value0)(_210.value1.value1.value1.value1.value1);
    };
};
var con5 = function (f) {
    return function (_209) {
        return f(_209.value0)(_209.value1.value0)(_209.value1.value1.value0)(_209.value1.value1.value1.value0)(_209.value1.value1.value1.value1);
    };
};
var con4 = function (f) {
    return function (_208) {
        return f(_208.value0)(_208.value1.value0)(_208.value1.value1.value0)(_208.value1.value1.value1);
    };
};
var con3 = function (f) {
    return function (_207) {
        return f(_207.value0)(_207.value1.value0)(_207.value1.value1);
    };
};
var con2 = function (f) {
    return function (_206) {
        return f(_206.value0)(_206.value1);
    };
};
var con10 = function (f) {
    return function (_214) {
        return f(_214.value0)(_214.value1.value0)(_214.value1.value1.value0)(_214.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value1.value1.value1.value1.value0)(_214.value1.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
module.exports = {
    "/\\": $div$bslash, 
    con10: con10, 
    con2: con2, 
    con3: con3, 
    con4: con4, 
    con5: con5, 
    con6: con6, 
    con7: con7, 
    con8: con8, 
    con9: con9
};
