// Data.Monoid.Sum/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Sum = {
    create: function (value) {
        return value;
    }
};
var showSum = new Prelude.Show(function (_130) {
    return "Sum (" + (Prelude.show(Prelude.showNumber)(_130) + ")");
});
var semigroupSum = new Prelude.Semigroup(function (_131) {
    return function (_132) {
        return _131 + _132;
    };
});
var runSum = function (_123) {
    return _123;
};
var monoidSum = new Data_Monoid.Monoid(function () {
    return semigroupSum;
}, 0);
var eqSum = new Prelude.Eq(function (_126) {
    return function (_127) {
        return _126 !== _127;
    };
}, function (_124) {
    return function (_125) {
        return _124 === _125;
    };
});
var ordSum = new Prelude.Ord(function () {
    return eqSum;
}, function (_128) {
    return function (_129) {
        return Prelude.compare(Prelude.ordNumber)(_128)(_129);
    };
});
module.exports = {
    Sum: Sum, 
    eqSum: eqSum, 
    monoidSum: monoidSum, 
    ordSum: ordSum, 
    runSum: runSum, 
    semigroupSum: semigroupSum, 
    showSum: showSum
};
