// Data.Monoid.Sum/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Sum = {
    create: function (value) {
        return value;
    }
};
var showSum = new Prelude.Show(function (_180) {
    return "Sum (" + (Prelude.show(Prelude.showNumber)(_180) + ")");
});
var semigroupSum = new Prelude.Semigroup(function (_181) {
    return function (_182) {
        return _181 + _182;
    };
});
var runSum = function (_173) {
    return _173;
};
var monoidSum = new Data_Monoid.Monoid(function () {
    return semigroupSum;
}, 0);
var eqSum = new Prelude.Eq(function (_176) {
    return function (_177) {
        return _176 !== _177;
    };
}, function (_174) {
    return function (_175) {
        return _174 === _175;
    };
});
var ordSum = new Prelude.Ord(function () {
    return eqSum;
}, function (_178) {
    return function (_179) {
        return Prelude.compare(Prelude.ordNumber)(_178)(_179);
    };
});
module.exports = {
    Sum: Sum, 
    eqSum: eqSum, 
    monoidSum: monoidSum, 
    ordSum: ordSum, 
    runSum: runSum, 
    semigroupSum: semigroupSum, 
    showSum: showSum
};
