// Data.Monoid.Last/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var Last = {
    create: function (value) {
        return value;
    }
};
var showLast = function (__dict_Show_0) {
    return new Prelude.Show(function (_150) {
        return "Last (" + (Prelude.show(Data_Maybe.showMaybe(__dict_Show_0))(_150) + ")");
    });
};
var semigroupLast = new Prelude.Semigroup(function (_151) {
    return function (_152) {
        if (_152 instanceof Data_Maybe.Just) {
            return _152;
        };
        if (_152 instanceof Data_Maybe.Nothing) {
            return _151;
        };
        throw new Error("Failed pattern match");
    };
});
var runLast = function (_143) {
    return _143;
};
var monoidLast = new Data_Monoid.Monoid(function () {
    return semigroupLast;
}, Data_Maybe.Nothing.value);
var eqLast = function (__dict_Eq_2) {
    return new Prelude.Eq(function (_146) {
        return function (_147) {
            return Prelude["/="](Data_Maybe.eqMaybe(__dict_Eq_2))(_146)(_147);
        };
    }, function (_144) {
        return function (_145) {
            return Prelude["=="](Data_Maybe.eqMaybe(__dict_Eq_2))(_144)(_145);
        };
    });
};
var ordLast = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqLast(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_148) {
        return function (_149) {
            return Prelude.compare(Data_Maybe.ordMaybe(__dict_Ord_1))(_148)(_149);
        };
    });
};
module.exports = {
    Last: Last, 
    eqLast: eqLast, 
    monoidLast: monoidLast, 
    ordLast: ordLast, 
    runLast: runLast, 
    semigroupLast: semigroupLast, 
    showLast: showLast
};
