// Data.Maybe/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Extend = require("Control.Extend");
var Control_Alt = require("Control.Alt");
var Control_Plus = require("Control.Plus");
var Control_Alternative = require("Control.Alternative");
var Control_MonadPlus = require("Control.MonadPlus");
function Nothing() {

};
Nothing.value = new Nothing();
function Just(value0) {
    this.value0 = value0;
};
Just.create = function (value0) {
    return new Just(value0);
};
var showMaybe = function (__dict_Show_0) {
    return new Prelude.Show(function (_74) {
        if (_74 instanceof Just) {
            return "Just (" + (Prelude.show(__dict_Show_0)(_74.value0) + ")");
        };
        if (_74 instanceof Nothing) {
            return "Nothing";
        };
        throw new Error("Failed pattern match");
    });
};
var semigroupMaybe = function (__dict_Semigroup_1) {
    return new Prelude.Semigroup(function (_72) {
        return function (_73) {
            if (_72 instanceof Nothing) {
                return _73;
            };
            if (_73 instanceof Nothing) {
                return _72;
            };
            if (_72 instanceof Just && _73 instanceof Just) {
                return new Just(Prelude["<>"](__dict_Semigroup_1)(_72.value0)(_73.value0));
            };
            throw new Error("Failed pattern match");
        };
    });
};
var maybe = function (_59) {
    return function (_60) {
        return function (_61) {
            if (_61 instanceof Nothing) {
                return _59;
            };
            if (_61 instanceof Just) {
                return _60(_61.value0);
            };
            throw new Error("Failed pattern match");
        };
    };
};
var isNothing = maybe(true)(Prelude["const"](false));
var isJust = maybe(false)(Prelude["const"](true));
var functorMaybe = new Prelude.Functor(function (_62) {
    return function (_63) {
        if (_63 instanceof Just) {
            return new Just(_62(_63.value0));
        };
        return Nothing.value;
    };
});
var fromMaybe = function (a) {
    return maybe(a)(Prelude.id(Prelude.categoryArr));
};
var extendMaybe = new Control_Extend.Extend(function (_70) {
    return function (_71) {
        if (_71 instanceof Nothing) {
            return Nothing.value;
        };
        return Just.create(_70(_71));
    };
}, function () {
    return functorMaybe;
});
var eqMaybe = function (__dict_Eq_3) {
    return new Prelude.Eq(function (a) {
        return function (b) {
            return !Prelude["=="](eqMaybe(__dict_Eq_3))(a)(b);
        };
    }, function (_75) {
        return function (_76) {
            if (_75 instanceof Nothing && _76 instanceof Nothing) {
                return true;
            };
            if (_75 instanceof Just && _76 instanceof Just) {
                return Prelude["=="](__dict_Eq_3)(_75.value0)(_76.value0);
            };
            return false;
        };
    });
};
var ordMaybe = function (__dict_Ord_2) {
    return new Prelude.Ord(function () {
        return eqMaybe(__dict_Ord_2["__superclass_Prelude.Eq_0"]());
    }, function (_77) {
        return function (_78) {
            if (_77 instanceof Just && _78 instanceof Just) {
                return Prelude.compare(__dict_Ord_2)(_77.value0)(_78.value0);
            };
            if (_77 instanceof Nothing && _78 instanceof Nothing) {
                return Prelude.EQ.value;
            };
            if (_77 instanceof Nothing) {
                return Prelude.LT.value;
            };
            if (_78 instanceof Nothing) {
                return Prelude.GT.value;
            };
            throw new Error("Failed pattern match");
        };
    });
};
var applyMaybe = new Prelude.Apply(function (_64) {
    return function (_65) {
        if (_64 instanceof Just) {
            return Prelude["<$>"](functorMaybe)(_64.value0)(_65);
        };
        if (_64 instanceof Nothing) {
            return Nothing.value;
        };
        throw new Error("Failed pattern match");
    };
}, function () {
    return functorMaybe;
});
var bindMaybe = new Prelude.Bind(function (_68) {
    return function (_69) {
        if (_68 instanceof Just) {
            return _69(_68.value0);
        };
        if (_68 instanceof Nothing) {
            return Nothing.value;
        };
        throw new Error("Failed pattern match");
    };
}, function () {
    return applyMaybe;
});
var applicativeMaybe = new Prelude.Applicative(function () {
    return applyMaybe;
}, Just.create);
var monadMaybe = new Prelude.Monad(function () {
    return applicativeMaybe;
}, function () {
    return bindMaybe;
});
var altMaybe = new Control_Alt.Alt(function (_66) {
    return function (_67) {
        if (_66 instanceof Nothing) {
            return _67;
        };
        return _66;
    };
}, function () {
    return functorMaybe;
});
var plusMaybe = new Control_Plus.Plus(function () {
    return altMaybe;
}, Nothing.value);
var alternativeMaybe = new Control_Alternative.Alternative(function () {
    return plusMaybe;
}, function () {
    return applicativeMaybe;
});
var monadPlusMaybe = new Control_MonadPlus.MonadPlus(function () {
    return alternativeMaybe;
}, function () {
    return monadMaybe;
});
module.exports = {
    Just: Just, 
    Nothing: Nothing, 
    altMaybe: altMaybe, 
    alternativeMaybe: alternativeMaybe, 
    applicativeMaybe: applicativeMaybe, 
    applyMaybe: applyMaybe, 
    bindMaybe: bindMaybe, 
    eqMaybe: eqMaybe, 
    extendMaybe: extendMaybe, 
    fromMaybe: fromMaybe, 
    functorMaybe: functorMaybe, 
    isJust: isJust, 
    isNothing: isNothing, 
    maybe: maybe, 
    monadMaybe: monadMaybe, 
    monadPlusMaybe: monadPlusMaybe, 
    ordMaybe: ordMaybe, 
    plusMaybe: plusMaybe, 
    semigroupMaybe: semigroupMaybe, 
    showMaybe: showMaybe
};
