// Data.Array.ST/index.js
"use strict";
var Data_Function = require("Data.Function");
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");

  function runSTArray(f) {
    return f;
  };

  function emptySTArray() {
    return [];
  };

  function peekSTArrayImpl(arr, i, s, f) {
    return function() {
      var index = ~~i;
      if (0 <= index && index < arr.length) {
        return s(arr[index]);
      } else {
        return f;
      }
    };
  };

  function pokeSTArrayImpl(arr, i, a) {
    return function() {
      var index = ~~i;
      if (0 <= index && index <= arr.length) {
        arr[index] = a;
        return true;
      }
      return false;
    };
  };

  function pushSTArrayImpl(arr, a) {
    return function() {
      arr.push(a);
      return {};
    };
  };
var pushSTArray = function (arr) {
    return function (a) {
        return pushSTArrayImpl(arr, a);
    };
};
var pokeSTArray = function (arr) {
    return function (i) {
        return function (a) {
            return pokeSTArrayImpl(arr, i, a);
        };
    };
};
var peekSTArray = function (arr) {
    return function (i) {
        return peekSTArrayImpl(arr, i, Data_Maybe.Just.create, Data_Maybe.Nothing.value);
    };
};
module.exports = {
    emptySTArray: emptySTArray, 
    peekSTArray: peekSTArray, 
    pokeSTArray: pokeSTArray, 
    pushSTArray: pushSTArray, 
    runSTArray: runSTArray
};
