// Data.Tuple.Nested/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Tuple = require("Data.Tuple");
var $div$bslash = function (a) {
    return function (b) {
        return new Data_Tuple.Tuple(a, b);
    };
};
var con9 = function (f) {
    return function (_55) {
        return f(_55.value0)(_55.value1.value0)(_55.value1.value1.value0)(_55.value1.value1.value1.value0)(_55.value1.value1.value1.value1.value0)(_55.value1.value1.value1.value1.value1.value0)(_55.value1.value1.value1.value1.value1.value1.value0)(_55.value1.value1.value1.value1.value1.value1.value1.value0)(_55.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con8 = function (f) {
    return function (_54) {
        return f(_54.value0)(_54.value1.value0)(_54.value1.value1.value0)(_54.value1.value1.value1.value0)(_54.value1.value1.value1.value1.value0)(_54.value1.value1.value1.value1.value1.value0)(_54.value1.value1.value1.value1.value1.value1.value0)(_54.value1.value1.value1.value1.value1.value1.value1);
    };
};
var con7 = function (f) {
    return function (_53) {
        return f(_53.value0)(_53.value1.value0)(_53.value1.value1.value0)(_53.value1.value1.value1.value0)(_53.value1.value1.value1.value1.value0)(_53.value1.value1.value1.value1.value1.value0)(_53.value1.value1.value1.value1.value1.value1);
    };
};
var con6 = function (f) {
    return function (_52) {
        return f(_52.value0)(_52.value1.value0)(_52.value1.value1.value0)(_52.value1.value1.value1.value0)(_52.value1.value1.value1.value1.value0)(_52.value1.value1.value1.value1.value1);
    };
};
var con5 = function (f) {
    return function (_51) {
        return f(_51.value0)(_51.value1.value0)(_51.value1.value1.value0)(_51.value1.value1.value1.value0)(_51.value1.value1.value1.value1);
    };
};
var con4 = function (f) {
    return function (_50) {
        return f(_50.value0)(_50.value1.value0)(_50.value1.value1.value0)(_50.value1.value1.value1);
    };
};
var con3 = function (f) {
    return function (_49) {
        return f(_49.value0)(_49.value1.value0)(_49.value1.value1);
    };
};
var con2 = function (f) {
    return function (_48) {
        return f(_48.value0)(_48.value1);
    };
};
var con10 = function (f) {
    return function (_56) {
        return f(_56.value0)(_56.value1.value0)(_56.value1.value1.value0)(_56.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value1.value1.value1.value1.value0)(_56.value1.value1.value1.value1.value1.value1.value1.value1.value1);
    };
};
module.exports = {
    "/\\": $div$bslash, 
    con10: con10, 
    con2: con2, 
    con3: con3, 
    con4: con4, 
    con5: con5, 
    con6: con6, 
    con7: con7, 
    con8: con8, 
    con9: con9
};
