// Main/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Reactive = require("Control.Reactive");
var Utils = require("Utils");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Component = {
    create: function (value) {
        return value;
    }
};
var Componen = {
    create: function (value) {
        return value;
    }
};
var sub = function (_4) {
    return function (_5) {
        return Control_Reactive.subscribe(_5.output)(_4);
    };
};
var set = function (_2) {
    return function (_3) {
        return function __do() {
            Control_Reactive.writeRVar(_3.input)(_2)();
            return Prelude.unit;
        };
    };
};
var mkComponent = function (input) {
    return function (output) {
        return function __do() {
            var _1 = Control_Reactive.newRVar(input)();
            var _0 = Control_Reactive.newRVar(output)();
            return {
                input: _1, 
                output: _0
            };
        };
    };
};
var main = Utils.whisper("foo");
var init = function (_6) {
    return function (_7) {
        return function __do() {
            Control_Reactive.subscribe(_7.input)(_6)();
            return _7;
        };
    };
};
module.exports = {
    Componen: Componen, 
    Component: Component, 
    init: init, 
    main: main, 
    mkComponent: mkComponent, 
    set: set, 
    sub: sub
};
