// Data.Traversable/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Array = require("Data.Array");
var Data_Tuple = require("Data.Tuple");
var Data_Foldable = require("Data.Foldable");
var Data_Maybe = require("Data.Maybe");
var Data_Either = require("Data.Either");
var StateR = {
    create: function (value) {
        return value;
    }
};
var StateL = {
    create: function (value) {
        return value;
    }
};
function Traversable(__superclass_Data$dotFoldable$dotFoldable_1, __superclass_Prelude$dotFunctor_0, sequence, traverse) {
    this["__superclass_Data.Foldable.Foldable_1"] = __superclass_Data$dotFoldable$dotFoldable_1;
    this["__superclass_Prelude.Functor_0"] = __superclass_Prelude$dotFunctor_0;
    this.sequence = sequence;
    this.traverse = traverse;
};
var traverse = function (dict) {
    return dict.traverse;
};
var traversableTuple = new Traversable(function () {
    return Data_Foldable.foldableTuple;
}, function () {
    return Data_Tuple.functorTuple;
}, function (__dict_Applicative_1) {
    return function (_98) {
        return Prelude["<$>"]((__dict_Applicative_1["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Tuple.Tuple.create(_98.value0))(_98.value1);
    };
}, function (__dict_Applicative_0) {
    return function (_96) {
        return function (_97) {
            return Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Tuple.Tuple.create(_97.value0))(_96(_97.value1));
        };
    };
});
var traversableMaybe = new Traversable(function () {
    return Data_Foldable.foldableMaybe;
}, function () {
    return Data_Maybe.functorMaybe;
}, function (__dict_Applicative_3) {
    return function (_95) {
        if (_95 instanceof Data_Maybe.Nothing) {
            return Prelude.pure(__dict_Applicative_3)(Data_Maybe.Nothing.value);
        };
        if (_95 instanceof Data_Maybe.Just) {
            return Prelude["<$>"]((__dict_Applicative_3["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Maybe.Just.create)(_95.value0);
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_2) {
    return function (_93) {
        return function (_94) {
            if (_94 instanceof Data_Maybe.Nothing) {
                return Prelude.pure(__dict_Applicative_2)(Data_Maybe.Nothing.value);
            };
            if (_94 instanceof Data_Maybe.Just) {
                return Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Maybe.Just.create)(_93(_94.value0));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableEither = new Traversable(function () {
    return Data_Foldable.foldableEither;
}, function () {
    return Data_Either.functorEither;
}, function (__dict_Applicative_5) {
    return function (_92) {
        if (_92 instanceof Data_Either.Left) {
            return Prelude.pure(__dict_Applicative_5)(new Data_Either.Left(_92.value0));
        };
        if (_92 instanceof Data_Either.Right) {
            return Prelude["<$>"]((__dict_Applicative_5["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Either.Right.create)(_92.value0);
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_4) {
    return function (_90) {
        return function (_91) {
            if (_91 instanceof Data_Either.Left) {
                return Prelude.pure(__dict_Applicative_4)(new Data_Either.Left(_91.value0));
            };
            if (_91 instanceof Data_Either.Right) {
                return Prelude["<$>"]((__dict_Applicative_4["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Either.Right.create)(_90(_91.value0));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var stateR = function (_86) {
    return _86;
};
var stateL = function (_85) {
    return _85;
};
var sequence = function (dict) {
    return dict.sequence;
};
var traversableArray = new Traversable(function () {
    return Data_Foldable.foldableArray;
}, function () {
    return Data_Array.functorArray;
}, function (__dict_Applicative_7) {
    return function (_89) {
        if (_89.length === 0) {
            return Prelude.pure(__dict_Applicative_7)([  ]);
        };
        if (_89.length >= 1) {
            var _416 = _89.slice(1);
            return Prelude["<*>"](__dict_Applicative_7["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_7["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude[":"])(_89[0]))(sequence(traversableArray)(__dict_Applicative_7)(_416));
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_6) {
    return function (_87) {
        return function (_88) {
            if (_88.length === 0) {
                return Prelude.pure(__dict_Applicative_6)([  ]);
            };
            if (_88.length >= 1) {
                var _420 = _88.slice(1);
                return Prelude["<*>"](__dict_Applicative_6["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_6["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude[":"])(_87(_88[0])))(traverse(traversableArray)(__dict_Applicative_6)(_87)(_420));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var zipWithA = function (__dict_Applicative_8) {
    return function (f) {
        return function (xs) {
            return function (ys) {
                return sequence(traversableArray)(__dict_Applicative_8)(Data_Array.zipWith(f)(xs)(ys));
            };
        };
    };
};
var functorStateR = new Prelude.Functor(function (f) {
    return function (k) {
        return StateR.create(function (s) {
            var _421 = stateR(k)(s);
            return new Data_Tuple.Tuple(_421.value0, f(_421.value1));
        });
    };
});
var functorStateL = new Prelude.Functor(function (f) {
    return function (k) {
        return StateL.create(function (s) {
            var _424 = stateL(k)(s);
            return new Data_Tuple.Tuple(_424.value0, f(_424.value1));
        });
    };
});
var $$for = function (__dict_Applicative_13) {
    return function (__dict_Traversable_14) {
        return function (x) {
            return function (f) {
                return traverse(__dict_Traversable_14)(__dict_Applicative_13)(f)(x);
            };
        };
    };
};
var applyStateR = new Prelude.Apply(function (f) {
    return function (x) {
        return StateR.create(function (s) {
            var _427 = stateR(x)(s);
            var _428 = stateR(f)(_427.value0);
            return new Data_Tuple.Tuple(_428.value0, _428.value1(_427.value1));
        });
    };
}, function () {
    return functorStateR;
});
var applyStateL = new Prelude.Apply(function (f) {
    return function (x) {
        return StateL.create(function (s) {
            var _433 = stateL(f)(s);
            var _434 = stateL(x)(_433.value0);
            return new Data_Tuple.Tuple(_434.value0, _433.value1(_434.value1));
        });
    };
}, function () {
    return functorStateL;
});
var applicativeStateR = new Prelude.Applicative(function () {
    return applyStateR;
}, function (a) {
    return StateR.create(function (s) {
        return new Data_Tuple.Tuple(s, a);
    });
});
var mapAccumR = function (__dict_Traversable_9) {
    return function (f) {
        return function (s0) {
            return function (xs) {
                return stateR(traverse(__dict_Traversable_9)(applicativeStateR)(function (a) {
                    return StateR.create(function (s) {
                        return f(s)(a);
                    });
                })(xs))(s0);
            };
        };
    };
};
var scanr = function (__dict_Traversable_10) {
    return function (f) {
        return function (b0) {
            return function (xs) {
                return Data_Tuple.snd(mapAccumR(__dict_Traversable_10)(function (b) {
                    return function (a) {
                        var b$prime = f(a)(b);
                        return new Data_Tuple.Tuple(b$prime, b$prime);
                    };
                })(b0)(xs));
            };
        };
    };
};
var applicativeStateL = new Prelude.Applicative(function () {
    return applyStateL;
}, function (a) {
    return StateL.create(function (s) {
        return new Data_Tuple.Tuple(s, a);
    });
});
var mapAccumL = function (__dict_Traversable_11) {
    return function (f) {
        return function (s0) {
            return function (xs) {
                return stateL(traverse(__dict_Traversable_11)(applicativeStateL)(function (a) {
                    return StateL.create(function (s) {
                        return f(s)(a);
                    });
                })(xs))(s0);
            };
        };
    };
};
var scanl = function (__dict_Traversable_12) {
    return function (f) {
        return function (b0) {
            return function (xs) {
                return Data_Tuple.snd(mapAccumL(__dict_Traversable_12)(function (b) {
                    return function (a) {
                        var b$prime = f(b)(a);
                        return new Data_Tuple.Tuple(b$prime, b$prime);
                    };
                })(b0)(xs));
            };
        };
    };
};
module.exports = {
    Traversable: Traversable, 
    "for": $$for, 
    mapAccumL: mapAccumL, 
    mapAccumR: mapAccumR, 
    scanl: scanl, 
    scanr: scanr, 
    sequence: sequence, 
    traversableArray: traversableArray, 
    traversableEither: traversableEither, 
    traversableMaybe: traversableMaybe, 
    traversableTuple: traversableTuple, 
    traverse: traverse, 
    zipWithA: zipWithA
};
