// Data.Monoid.First/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var First = {
    create: function (value) {
        return value;
    }
};
var showFirst = function (__dict_Show_0) {
    return new Prelude.Show(function (_58) {
        return "First (" + (Prelude.show(Data_Maybe.showMaybe(__dict_Show_0))(_58) + ")");
    });
};
var semigroupFirst = new Prelude.Semigroup(function (_59) {
    return function (_60) {
        if (_59 instanceof Data_Maybe.Just) {
            return _59;
        };
        return _60;
    };
});
var runFirst = function (_51) {
    return _51;
};
var monoidFirst = new Data_Monoid.Monoid(function () {
    return semigroupFirst;
}, Data_Maybe.Nothing.value);
var eqFirst = function (__dict_Eq_2) {
    return new Prelude.Eq(function (_54) {
        return function (_55) {
            return Prelude["/="](Data_Maybe.eqMaybe(__dict_Eq_2))(_54)(_55);
        };
    }, function (_52) {
        return function (_53) {
            return Prelude["=="](Data_Maybe.eqMaybe(__dict_Eq_2))(_52)(_53);
        };
    });
};
var ordFirst = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqFirst(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_56) {
        return function (_57) {
            return Prelude.compare(Data_Maybe.ordMaybe(__dict_Ord_1))(_56)(_57);
        };
    });
};
module.exports = {
    First: First, 
    eqFirst: eqFirst, 
    monoidFirst: monoidFirst, 
    ordFirst: ordFirst, 
    runFirst: runFirst, 
    semigroupFirst: semigroupFirst, 
    showFirst: showFirst
};
