// Data.Either/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Alt = require("Control.Alt");
function Left(value0) {
    this.value0 = value0;
};
Left.create = function (value0) {
    return new Left(value0);
};
function Right(value0) {
    this.value0 = value0;
};
Right.create = function (value0) {
    return new Right(value0);
};
var showEither = function (__dict_Show_0) {
    return function (__dict_Show_1) {
        return new Prelude.Show(function (_20) {
            if (_20 instanceof Left) {
                return "Left (" + (Prelude.show(__dict_Show_0)(_20.value0) + ")");
            };
            if (_20 instanceof Right) {
                return "Right (" + (Prelude.show(__dict_Show_1)(_20.value0) + ")");
            };
            throw new Error("Failed pattern match");
        });
    };
};
var functorEither = new Prelude.Functor(function (_14) {
    return function (_15) {
        if (_15 instanceof Left) {
            return new Left(_15.value0);
        };
        if (_15 instanceof Right) {
            return new Right(_14(_15.value0));
        };
        throw new Error("Failed pattern match");
    };
});
var eqEither = function (__dict_Eq_4) {
    return function (__dict_Eq_5) {
        return new Prelude.Eq(function (a) {
            return function (b) {
                return !Prelude["=="](eqEither(__dict_Eq_4)(__dict_Eq_5))(a)(b);
            };
        }, function (_21) {
            return function (_22) {
                if (_21 instanceof Left && _22 instanceof Left) {
                    return Prelude["=="](__dict_Eq_4)(_21.value0)(_22.value0);
                };
                if (_21 instanceof Right && _22 instanceof Right) {
                    return Prelude["=="](__dict_Eq_5)(_21.value0)(_22.value0);
                };
                return false;
            };
        });
    };
};
var ordEither = function (__dict_Ord_2) {
    return function (__dict_Ord_3) {
        return new Prelude.Ord(function () {
            return eqEither(__dict_Ord_2["__superclass_Prelude.Eq_0"]())(__dict_Ord_3["__superclass_Prelude.Eq_0"]());
        }, function (_23) {
            return function (_24) {
                if (_23 instanceof Left && _24 instanceof Left) {
                    return Prelude.compare(__dict_Ord_2)(_23.value0)(_24.value0);
                };
                if (_23 instanceof Right && _24 instanceof Right) {
                    return Prelude.compare(__dict_Ord_3)(_23.value0)(_24.value0);
                };
                if (_23 instanceof Left) {
                    return Prelude.LT.value;
                };
                if (_24 instanceof Left) {
                    return Prelude.GT.value;
                };
                throw new Error("Failed pattern match");
            };
        });
    };
};
var either = function (_11) {
    return function (_12) {
        return function (_13) {
            if (_13 instanceof Left) {
                return _11(_13.value0);
            };
            if (_13 instanceof Right) {
                return _12(_13.value0);
            };
            throw new Error("Failed pattern match");
        };
    };
};
var isLeft = either(Prelude["const"](true))(Prelude["const"](false));
var isRight = either(Prelude["const"](false))(Prelude["const"](true));
var applyEither = new Prelude.Apply(function (_16) {
    return function (_17) {
        if (_16 instanceof Left) {
            return new Left(_16.value0);
        };
        if (_16 instanceof Right) {
            return Prelude["<$>"](functorEither)(_16.value0)(_17);
        };
        throw new Error("Failed pattern match");
    };
}, function () {
    return functorEither;
});
var bindEither = new Prelude.Bind(either(function (e) {
    return function (_) {
        return new Left(e);
    };
})(function (a) {
    return function (f) {
        return f(a);
    };
}), function () {
    return applyEither;
});
var applicativeEither = new Prelude.Applicative(function () {
    return applyEither;
}, Right.create);
var monadEither = new Prelude.Monad(function () {
    return applicativeEither;
}, function () {
    return bindEither;
});
var altEither = new Control_Alt.Alt(function (_18) {
    return function (_19) {
        if (_18 instanceof Left) {
            return _19;
        };
        return _18;
    };
}, function () {
    return functorEither;
});
module.exports = {
    Left: Left, 
    Right: Right, 
    altEither: altEither, 
    applicativeEither: applicativeEither, 
    applyEither: applyEither, 
    bindEither: bindEither, 
    either: either, 
    eqEither: eqEither, 
    functorEither: functorEither, 
    isLeft: isLeft, 
    isRight: isRight, 
    monadEither: monadEither, 
    ordEither: ordEither, 
    showEither: showEither
};
