// Data.Monoid.Dual/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Dual = {
    create: function (value) {
        return value;
    }
};
var showDual = function (__dict_Show_0) {
    return new Prelude.Show(function (_157) {
        return "Dual (" + (Prelude.show(__dict_Show_0)(_157) + ")");
    });
};
var semigroupDual = function (__dict_Semigroup_1) {
    return new Prelude.Semigroup(function (_158) {
        return function (_159) {
            return Prelude["<>"](__dict_Semigroup_1)(_159)(_158);
        };
    });
};
var runDual = function (_150) {
    return _150;
};
var monoidDual = function (__dict_Monoid_3) {
    return new Data_Monoid.Monoid(function () {
        return semigroupDual(__dict_Monoid_3["__superclass_Prelude.Semigroup_0"]());
    }, Data_Monoid.mempty(__dict_Monoid_3));
};
var eqDual = function (__dict_Eq_4) {
    return new Prelude.Eq(function (_153) {
        return function (_154) {
            return Prelude["/="](__dict_Eq_4)(_153)(_154);
        };
    }, function (_151) {
        return function (_152) {
            return Prelude["=="](__dict_Eq_4)(_151)(_152);
        };
    });
};
var ordDual = function (__dict_Ord_2) {
    return new Prelude.Ord(function () {
        return eqDual(__dict_Ord_2["__superclass_Prelude.Eq_0"]());
    }, function (_155) {
        return function (_156) {
            return Prelude.compare(__dict_Ord_2)(_155)(_156);
        };
    });
};
module.exports = {
    Dual: Dual, 
    eqDual: eqDual, 
    monoidDual: monoidDual, 
    ordDual: ordDual, 
    runDual: runDual, 
    semigroupDual: semigroupDual, 
    showDual: showDual
};
