// Data.Monoid.Sum/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Sum = {
    create: function (value) {
        return value;
    }
};
var showSum = new Prelude.Show(function (_48) {
    return "Sum (" + (Prelude.show(Prelude.showNumber)(_48) + ")");
});
var semigroupSum = new Prelude.Semigroup(function (_49) {
    return function (_50) {
        return _49 + _50;
    };
});
var runSum = function (_41) {
    return _41;
};
var monoidSum = new Data_Monoid.Monoid(function () {
    return semigroupSum;
}, 0);
var eqSum = new Prelude.Eq(function (_44) {
    return function (_45) {
        return _44 !== _45;
    };
}, function (_42) {
    return function (_43) {
        return _42 === _43;
    };
});
var ordSum = new Prelude.Ord(function () {
    return eqSum;
}, function (_46) {
    return function (_47) {
        return Prelude.compare(Prelude.ordNumber)(_46)(_47);
    };
});
module.exports = {
    Sum: Sum, 
    eqSum: eqSum, 
    monoidSum: monoidSum, 
    ordSum: ordSum, 
    runSum: runSum, 
    semigroupSum: semigroupSum, 
    showSum: showSum
};
