// Data.Char/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Char = {
    create: function (value) {
        return value;
    }
};

    function toCharCode(c) {
      return c.charCodeAt(0);
    }
    ;

    function fromCharCode(c) {
      return String.fromCharCode(c);
    }
    ;
var showChar = new Prelude.Show(function (_55) {
    return "Char " + Prelude.show(Prelude.showString)(_55);
});
var eqChar = new Prelude.Eq(function (a) {
    return function (b) {
        return !Prelude["=="](eqChar)(a)(b);
    };
}, function (_51) {
    return function (_52) {
        return _51 === _52;
    };
});
var ordChar = new Prelude.Ord(function () {
    return eqChar;
}, function (_53) {
    return function (_54) {
        return Prelude.compare(Prelude.ordString)(_53)(_54);
    };
});
var charString = function (_50) {
    return _50;
};
module.exports = {
    charString: charString, 
    eqChar: eqChar, 
    fromCharCode: fromCharCode, 
    ordChar: ordChar, 
    showChar: showChar, 
    toCharCode: toCharCode
};
