// Data.Monoid.Product/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Product = {
    create: function (value) {
        return value;
    }
};
var showProduct = new Prelude.Show(function (_120) {
    return "Product (" + (Prelude.show(Prelude.showNumber)(_120) + ")");
});
var semigroupProduct = new Prelude.Semigroup(function (_121) {
    return function (_122) {
        return _121 * _122;
    };
});
var runProduct = function (_113) {
    return _113;
};
var monoidProduct = new Data_Monoid.Monoid(function () {
    return semigroupProduct;
}, 1);
var eqProduct = new Prelude.Eq(function (_116) {
    return function (_117) {
        return _116 !== _117;
    };
}, function (_114) {
    return function (_115) {
        return _114 === _115;
    };
});
var ordProduct = new Prelude.Ord(function () {
    return eqProduct;
}, function (_118) {
    return function (_119) {
        return Prelude.compare(Prelude.ordNumber)(_118)(_119);
    };
});
module.exports = {
    Product: Product, 
    eqProduct: eqProduct, 
    monoidProduct: monoidProduct, 
    ordProduct: ordProduct, 
    runProduct: runProduct, 
    semigroupProduct: semigroupProduct, 
    showProduct: showProduct
};
