// Data.Monoid.Product/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Product = {
    create: function (value) {
        return value;
    }
};
var showProduct = new Prelude.Show(function (_170) {
    return "Product (" + (Prelude.show(Prelude.showNumber)(_170) + ")");
});
var semigroupProduct = new Prelude.Semigroup(function (_171) {
    return function (_172) {
        return _171 * _172;
    };
});
var runProduct = function (_163) {
    return _163;
};
var monoidProduct = new Data_Monoid.Monoid(function () {
    return semigroupProduct;
}, 1);
var eqProduct = new Prelude.Eq(function (_166) {
    return function (_167) {
        return _166 !== _167;
    };
}, function (_164) {
    return function (_165) {
        return _164 === _165;
    };
});
var ordProduct = new Prelude.Ord(function () {
    return eqProduct;
}, function (_168) {
    return function (_169) {
        return Prelude.compare(Prelude.ordNumber)(_168)(_169);
    };
});
module.exports = {
    Product: Product, 
    eqProduct: eqProduct, 
    monoidProduct: monoidProduct, 
    ordProduct: ordProduct, 
    runProduct: runProduct, 
    semigroupProduct: semigroupProduct, 
    showProduct: showProduct
};
