// Data.Monoid.Last/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var Last = {
    create: function (value) {
        return value;
    }
};
var showLast = function (__dict_Show_0) {
    return new Prelude.Show(function (_68) {
        return "Last (" + (Prelude.show(Data_Maybe.showMaybe(__dict_Show_0))(_68) + ")");
    });
};
var semigroupLast = new Prelude.Semigroup(function (_69) {
    return function (_70) {
        if (_70 instanceof Data_Maybe.Just) {
            return _70;
        };
        if (_70 instanceof Data_Maybe.Nothing) {
            return _69;
        };
        throw new Error("Failed pattern match");
    };
});
var runLast = function (_61) {
    return _61;
};
var monoidLast = new Data_Monoid.Monoid(function () {
    return semigroupLast;
}, Data_Maybe.Nothing.value);
var eqLast = function (__dict_Eq_2) {
    return new Prelude.Eq(function (_64) {
        return function (_65) {
            return Prelude["/="](Data_Maybe.eqMaybe(__dict_Eq_2))(_64)(_65);
        };
    }, function (_62) {
        return function (_63) {
            return Prelude["=="](Data_Maybe.eqMaybe(__dict_Eq_2))(_62)(_63);
        };
    });
};
var ordLast = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqLast(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_66) {
        return function (_67) {
            return Prelude.compare(Data_Maybe.ordMaybe(__dict_Ord_1))(_66)(_67);
        };
    });
};
module.exports = {
    Last: Last, 
    eqLast: eqLast, 
    monoidLast: monoidLast, 
    ordLast: ordLast, 
    runLast: runLast, 
    semigroupLast: semigroupLast, 
    showLast: showLast
};
