// Data.Char/index.js
"use strict";
var Prelude = require("Prelude");
var Char = {
    create: function (value) {
        return value;
    }
};

    function toCharCode(c) {
      return c.charCodeAt(0);
    }
    ;

    function fromCharCode(c) {
      return String.fromCharCode(c);
    }
    ;
var showChar = new Prelude.Show(function (_10) {
    return "Char " + Prelude.show(Prelude.showString)(_10);
});
var eqChar = new Prelude.Eq(function (a) {
    return function (b) {
        return !Prelude["=="](eqChar)(a)(b);
    };
}, function (_6) {
    return function (_7) {
        return _6 === _7;
    };
});
var ordChar = new Prelude.Ord(function () {
    return eqChar;
}, function (_8) {
    return function (_9) {
        return Prelude.compare(Prelude.ordString)(_8)(_9);
    };
});
var charString = function (_5) {
    return _5;
};
module.exports = {
    charString: charString, 
    eqChar: eqChar, 
    fromCharCode: fromCharCode, 
    ordChar: ordChar, 
    showChar: showChar, 
    toCharCode: toCharCode
};
