// Data.Monoid.Product/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Product = {
    create: function (value) {
        return value;
    }
};
var showProduct = new Prelude.Show(function (_38) {
    return "Product (" + (Prelude.show(Prelude.showNumber)(_38) + ")");
});
var semigroupProduct = new Prelude.Semigroup(function (_39) {
    return function (_40) {
        return _39 * _40;
    };
});
var runProduct = function (_31) {
    return _31;
};
var monoidProduct = new Data_Monoid.Monoid(function () {
    return semigroupProduct;
}, 1);
var eqProduct = new Prelude.Eq(function (_34) {
    return function (_35) {
        return _34 !== _35;
    };
}, function (_32) {
    return function (_33) {
        return _32 === _33;
    };
});
var ordProduct = new Prelude.Ord(function () {
    return eqProduct;
}, function (_36) {
    return function (_37) {
        return Prelude.compare(Prelude.ordNumber)(_36)(_37);
    };
});
module.exports = {
    Product: Product, 
    eqProduct: eqProduct, 
    monoidProduct: monoidProduct, 
    ordProduct: ordProduct, 
    runProduct: runProduct, 
    semigroupProduct: semigroupProduct, 
    showProduct: showProduct
};
