// Control.Monad.JQuery/index.js
"use strict";
var Prelude = require("Prelude");

  function ready(func) {
    return function () {
      jQuery(document).ready(func);
    };
  }
  ;

  function select(selector) {
    return function () {
      return jQuery(selector);
    };
  }
  ;

  function find(selector) {
    return function(ob) {
      return function () {
        return ob.find(selector);
      };
    };
  }
  ;

  function parent(ob) {
    return function () {
      return ob.parent();
    };
  }
  ;

  function closest(selector) {
    return function(ob) {
      return function () {
        return ob.closest(selector);
      };
    };
  }
  ;

  function create(html) {
    return function () {
      return jQuery(html);
    };
  }
  ;

  function setAttr(attr) {
    return function(val) {
      return function(ob) {
        return function () {
          return ob.attr(attr, val);
        };
      };
    };
  }
  ;

  function attr(attrs) {
    return function(ob) {
      return function () {
        return ob.attr(attrs);
      };
    };
  }
  ;

  function css(props) {
    return function(ob) {
      return function () {
        return ob.css(props);
      };
    };
  }
  ;

  function hasClass(cls) {
    return function(ob) {
      return function () {
        return ob.hasClass(cls);
      };
    };
  }
  ;

  function addClass(cls) {
    return function(ob) {
      return function () {
        return ob.addClass(cls);
      };
    };
  }
  ;

  function removeClass(cls) {
    return function(ob) {
      return function () {
        return ob.removeClass(cls);
      };
    };
  }
  ;

  function setProp(p) {
    return function(val) {
      return function(ob) {
        return function () {
          return ob.prop(p, val);
        };
      };
    };
  }
  ;

  function getProp(p) {
    return function(ob) {
      return function () {
        return ob.prop(p);
      };
    };
  }
  ;

  function append(ob1) {
    return function(ob) {
      return function () {
        return ob.append(ob1);
      };
    };
  }
  ;

  function appendAtIndex(i) {
    return function(ob1) {
      return function(ob) {
        return function () {
          var children = ob.children();
          if (children.length > 0) {
            if (i <= 0) {
               ob.prepend(ob1);
            } else if (i >= children.length) {
               ob.append(ob1);
            } else {
               ob1.insertBefore(jQuery(children[i]));
            }
            return ob;
          } else {
            return ob.append(ob1);
          }
        };
      };
    };
  }
  ;

  function remove(ob) {
    return function () {
      return ob.remove();
    };
  }
  ;

  function clear(ob) {
    return function () {
      return ob.empty();
    };
  }
  ;

  function before(ob) {
    return function(ob1) {
      return function () {
        return ob1.before(ob);
      };
    };
  }
  ;

  function appendText(s) {
    return function(ob) {
      return function () {
        return ob.append(s);
      };
    };
  }
  ;

  function body() {
    return jQuery(document.body);
  }
  ;

  function getText(ob) {
    return function() {
      return ob.text();
    };
  }
  ;

  function setText(text) {
    return function(ob) {
      return function() {
        return ob.text(text);
      };
    };
  }
  ;

  function getValue(ob) {
    return function() {
      return ob.val();
    };
  }
  ;

  function setValue(val) {
    return function(ob) {
      return function() {
        return ob.val(val);
      };
    };
  }
  ;

  function on(evt) {
    return function(act) {
      return function(ob) {
        return function() {
          return ob.on(evt, function(e) {
            act(e)(jQuery(this))();
          });
        };
      };
    };
  }
  ;

  function on$prime(evt) {
    return function(sel) {
      return function(act) {
        return function(ob) {
          return function() {
            return ob.on(evt, sel, function(e) {
              act(e)(jQuery(this))();
            });
          };
        };
      };
    };
  }
  ;

  function preventDefault(e) {
    return function() {
      e.preventDefault();
    };
  }
  ;

  function stopPropagation(e) {
    return function() {
      e.stopPropagation();
    };
  }
  ;

  function stopImmediatePropagation(e) {
    return function() {
      e.stopImmediatePropagation();
    };
  }
  ;
module.exports = {
    addClass: addClass, 
    append: append, 
    appendAtIndex: appendAtIndex, 
    appendText: appendText, 
    attr: attr, 
    before: before, 
    body: body, 
    clear: clear, 
    closest: closest, 
    create: create, 
    css: css, 
    find: find, 
    getProp: getProp, 
    getText: getText, 
    getValue: getValue, 
    hasClass: hasClass, 
    on: on, 
    "on'": on$prime, 
    parent: parent, 
    preventDefault: preventDefault, 
    ready: ready, 
    remove: remove, 
    removeClass: removeClass, 
    select: select, 
    setAttr: setAttr, 
    setProp: setProp, 
    setText: setText, 
    setValue: setValue, 
    stopImmediatePropagation: stopImmediatePropagation, 
    stopPropagation: stopPropagation
};
