// Data.Traversable/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Array = require("Data.Array");
var Data_Tuple = require("Data.Tuple");
var Data_Foldable = require("Data.Foldable");
var Data_Maybe = require("Data.Maybe");
var Data_Either = require("Data.Either");
var StateR = {
    create: function (value) {
        return value;
    }
};
var StateL = {
    create: function (value) {
        return value;
    }
};
function Traversable(__superclass_Data$dotFoldable$dotFoldable_1, __superclass_Prelude$dotFunctor_0, sequence, traverse) {
    this["__superclass_Data.Foldable.Foldable_1"] = __superclass_Data$dotFoldable$dotFoldable_1;
    this["__superclass_Prelude.Functor_0"] = __superclass_Prelude$dotFunctor_0;
    this.sequence = sequence;
    this.traverse = traverse;
};
var traverse = function (dict) {
    return dict.traverse;
};
var traversableTuple = new Traversable(function () {
    return Data_Foldable.foldableTuple;
}, function () {
    return Data_Tuple.functorTuple;
}, function (__dict_Applicative_1) {
    return function (_276) {
        return Prelude["<$>"]((__dict_Applicative_1["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Tuple.Tuple.create(_276.value0))(_276.value1);
    };
}, function (__dict_Applicative_0) {
    return function (_274) {
        return function (_275) {
            return Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Tuple.Tuple.create(_275.value0))(_274(_275.value1));
        };
    };
});
var traversableMaybe = new Traversable(function () {
    return Data_Foldable.foldableMaybe;
}, function () {
    return Data_Maybe.functorMaybe;
}, function (__dict_Applicative_3) {
    return function (_273) {
        if (_273 instanceof Data_Maybe.Nothing) {
            return Prelude.pure(__dict_Applicative_3)(Data_Maybe.Nothing.value);
        };
        if (_273 instanceof Data_Maybe.Just) {
            return Prelude["<$>"]((__dict_Applicative_3["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Maybe.Just.create)(_273.value0);
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_2) {
    return function (_271) {
        return function (_272) {
            if (_272 instanceof Data_Maybe.Nothing) {
                return Prelude.pure(__dict_Applicative_2)(Data_Maybe.Nothing.value);
            };
            if (_272 instanceof Data_Maybe.Just) {
                return Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Maybe.Just.create)(_271(_272.value0));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableEither = new Traversable(function () {
    return Data_Foldable.foldableEither;
}, function () {
    return Data_Either.functorEither;
}, function (__dict_Applicative_5) {
    return function (_270) {
        if (_270 instanceof Data_Either.Left) {
            return Prelude.pure(__dict_Applicative_5)(new Data_Either.Left(_270.value0));
        };
        if (_270 instanceof Data_Either.Right) {
            return Prelude["<$>"]((__dict_Applicative_5["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Either.Right.create)(_270.value0);
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_4) {
    return function (_268) {
        return function (_269) {
            if (_269 instanceof Data_Either.Left) {
                return Prelude.pure(__dict_Applicative_4)(new Data_Either.Left(_269.value0));
            };
            if (_269 instanceof Data_Either.Right) {
                return Prelude["<$>"]((__dict_Applicative_4["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Either.Right.create)(_268(_269.value0));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var stateR = function (_264) {
    return _264;
};
var stateL = function (_263) {
    return _263;
};
var sequence = function (dict) {
    return dict.sequence;
};
var traversableArray = new Traversable(function () {
    return Data_Foldable.foldableArray;
}, function () {
    return Data_Array.functorArray;
}, function (__dict_Applicative_7) {
    return function (_267) {
        if (_267.length === 0) {
            return Prelude.pure(__dict_Applicative_7)([  ]);
        };
        if (_267.length >= 1) {
            var _836 = _267.slice(1);
            return Prelude["<*>"](__dict_Applicative_7["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_7["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude[":"])(_267[0]))(sequence(traversableArray)(__dict_Applicative_7)(_836));
        };
        throw new Error("Failed pattern match");
    };
}, function (__dict_Applicative_6) {
    return function (_265) {
        return function (_266) {
            if (_266.length === 0) {
                return Prelude.pure(__dict_Applicative_6)([  ]);
            };
            if (_266.length >= 1) {
                var _840 = _266.slice(1);
                return Prelude["<*>"](__dict_Applicative_6["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_6["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude[":"])(_265(_266[0])))(traverse(traversableArray)(__dict_Applicative_6)(_265)(_840));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var zipWithA = function (__dict_Applicative_8) {
    return function (f) {
        return function (xs) {
            return function (ys) {
                return sequence(traversableArray)(__dict_Applicative_8)(Data_Array.zipWith(f)(xs)(ys));
            };
        };
    };
};
var functorStateR = new Prelude.Functor(function (f) {
    return function (k) {
        return StateR.create(function (s) {
            var _841 = stateR(k)(s);
            return new Data_Tuple.Tuple(_841.value0, f(_841.value1));
        });
    };
});
var functorStateL = new Prelude.Functor(function (f) {
    return function (k) {
        return StateL.create(function (s) {
            var _844 = stateL(k)(s);
            return new Data_Tuple.Tuple(_844.value0, f(_844.value1));
        });
    };
});
var $$for = function (__dict_Applicative_13) {
    return function (__dict_Traversable_14) {
        return function (x) {
            return function (f) {
                return traverse(__dict_Traversable_14)(__dict_Applicative_13)(f)(x);
            };
        };
    };
};
var applyStateR = new Prelude.Apply(function (f) {
    return function (x) {
        return StateR.create(function (s) {
            var _847 = stateR(x)(s);
            var _848 = stateR(f)(_847.value0);
            return new Data_Tuple.Tuple(_848.value0, _848.value1(_847.value1));
        });
    };
}, function () {
    return functorStateR;
});
var applyStateL = new Prelude.Apply(function (f) {
    return function (x) {
        return StateL.create(function (s) {
            var _853 = stateL(f)(s);
            var _854 = stateL(x)(_853.value0);
            return new Data_Tuple.Tuple(_854.value0, _853.value1(_854.value1));
        });
    };
}, function () {
    return functorStateL;
});
var applicativeStateR = new Prelude.Applicative(function () {
    return applyStateR;
}, function (a) {
    return StateR.create(function (s) {
        return new Data_Tuple.Tuple(s, a);
    });
});
var mapAccumR = function (__dict_Traversable_9) {
    return function (f) {
        return function (s0) {
            return function (xs) {
                return stateR(traverse(__dict_Traversable_9)(applicativeStateR)(function (a) {
                    return StateR.create(function (s) {
                        return f(s)(a);
                    });
                })(xs))(s0);
            };
        };
    };
};
var scanr = function (__dict_Traversable_10) {
    return function (f) {
        return function (b0) {
            return function (xs) {
                return Data_Tuple.snd(mapAccumR(__dict_Traversable_10)(function (b) {
                    return function (a) {
                        var b$prime = f(a)(b);
                        return new Data_Tuple.Tuple(b$prime, b$prime);
                    };
                })(b0)(xs));
            };
        };
    };
};
var applicativeStateL = new Prelude.Applicative(function () {
    return applyStateL;
}, function (a) {
    return StateL.create(function (s) {
        return new Data_Tuple.Tuple(s, a);
    });
});
var mapAccumL = function (__dict_Traversable_11) {
    return function (f) {
        return function (s0) {
            return function (xs) {
                return stateL(traverse(__dict_Traversable_11)(applicativeStateL)(function (a) {
                    return StateL.create(function (s) {
                        return f(s)(a);
                    });
                })(xs))(s0);
            };
        };
    };
};
var scanl = function (__dict_Traversable_12) {
    return function (f) {
        return function (b0) {
            return function (xs) {
                return Data_Tuple.snd(mapAccumL(__dict_Traversable_12)(function (b) {
                    return function (a) {
                        var b$prime = f(b)(a);
                        return new Data_Tuple.Tuple(b$prime, b$prime);
                    };
                })(b0)(xs));
            };
        };
    };
};
module.exports = {
    Traversable: Traversable, 
    "for": $$for, 
    mapAccumL: mapAccumL, 
    mapAccumR: mapAccumR, 
    scanl: scanl, 
    scanr: scanr, 
    sequence: sequence, 
    traversableArray: traversableArray, 
    traversableEither: traversableEither, 
    traversableMaybe: traversableMaybe, 
    traversableTuple: traversableTuple, 
    traverse: traverse, 
    zipWithA: zipWithA
};
