// Control.Reactive/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Eff_Ref_Unsafe = require("Control.Monad.Eff.Ref.Unsafe");
var Control_Monad_Eff_Ref = require("Control.Monad.Eff.Ref");
var Data_Monoid = require("Data.Monoid");
var Control_Monad_Eff = require("Control.Monad.Eff");
function Subscription(value0) {
    this.value0 = value0;
};
Subscription.create = function (value0) {
    return new Subscription(value0);
};
function Inserted(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
Inserted.create = function (value0) {
    return function (value1) {
        return new Inserted(value0, value1);
    };
};
function Updated(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
Updated.create = function (value0) {
    return function (value1) {
        return new Updated(value0, value1);
    };
};
function Removed(value0) {
    this.value0 = value0;
};
Removed.create = function (value0) {
    return new Removed(value0);
};
function Computed(value0) {
    this.value0 = value0;
};
Computed.create = function (value0) {
    return new Computed(value0);
};
function newRVar(value) {  return function() {    return (function () {      function RVar(value) {        var self = this;        self.value = value;        self.listeners = [];        self.subscribe = function (listener) {          this.listeners.push(listener);          return new Subscription(function() {            for (var i = 0; i < self.listeners.length; i++) {              if (self.listeners[i] === listener) {                self.listeners.splice(i, 1);                break;              }            }          });        };        self.update = function (value) {          self.value = value;          for (var i = 0; i < self.listeners.length; i++) {            self.listeners[i](value);          }        };      };      return new RVar(value);    })();  };};
function newRArray() {    return (function () {      function RArray() {        var self = this;        self.values = [];        self.listeners = [];        self.subscribe = function (listener) {          this.listeners.push(listener);          return Subscription(function() {            for (var i = 0; i < self.listeners.length; i++) {              if (self.listeners[i] === listener) {                self.listeners.splice(i, 1);                break;              }            }          });        };        self.insert = function (value, index) {          self.values.splice(index, 0, value);          for (var i = 0; i < self.listeners.length; i++) {            self.listeners[i](Inserted(value)(index));          }        };        self.remove = function (index) {          self.values.splice(index, 1);          for (var i = 0; i < self.listeners.length; i++) {            self.listeners[i](Removed(index));          }        };        self.update = function (value, index) {          self.values[index] = index;          for (var i = 0; i < self.listeners.length; i++) {            self.listeners[i](Updated(value)(index));          }        };      };      return new RArray();    })();};
function readRVar(ref) {  return function() {    return ref.value;  };};
function readRArray(arr) {  return function() {    return arr.values;  };};
function writeRVar(ref) {  return function (value) {    return function() {      ref.update(value);    };  };};
function peekRArray(arr) {  return function(i) {    return arr.values[i];  };};
function insertRArray(arr) {  return function (value) {    return function(index) {      return function() {        arr.insert(value, index);      };    };  };};
function removeRArray(arr) {  return function(index) {    return function() {      arr.remove(index);    };  };};
function updateRArray(arr) {  return function (value) {    return function(index) {      return function() {        arr.update(value, index);      };    };  };};
function subscribe(ref) {  return function(f) {    return function() {      return ref.subscribe(function(value) {        f(value)();      });    };  };};
function subscribeArray(arr) {  return function(f) {    return function() {      return arr.subscribe(function(value) {        f(value)();      });    };  };};
var toComputedArray = function (arr) {
    return new Computed({
        read: readRArray(arr), 
        subscribe: function (f) {
            return subscribeArray(arr)(function (_) {
                return Prelude[">>="](Control_Monad_Eff.bindEff)(readRArray(arr))(f);
            });
        }
    });
};
var toComputed = function (ref) {
    return new Computed({
        read: readRVar(ref), 
        subscribe: subscribe(ref)
    });
};
var subscribeComputed = function (_104) {
    return function (_105) {
        return _104.value0.subscribe(_105);
    };
};
var showArrayChange = function (__dict_Show_0) {
    return new Prelude.Show(function (_108) {
        if (_108 instanceof Inserted) {
            return "Inserted " + (Prelude.show(__dict_Show_0)(_108.value0) + (" at " + Prelude.show(Prelude.showNumber)(_108.value1)));
        };
        if (_108 instanceof Updated) {
            return "Updated " + (Prelude.show(Prelude.showNumber)(_108.value1) + (" to " + Prelude.show(__dict_Show_0)(_108.value0)));
        };
        if (_108 instanceof Removed) {
            return "Removed at index " + Prelude.show(Prelude.showNumber)(_108.value0);
        };
        throw new Error("Failed pattern match");
    });
};
var semigroupSubscription = new Prelude.Semigroup(function (_106) {
    return function (_107) {
        return new Subscription(function __do() {
            _106.value0();
            return _107.value0();
        });
    };
});
var readComputed = function (_103) {
    return _103.value0.read;
};
var monoidSubscription = new Data_Monoid.Monoid(function () {
    return semigroupSubscription;
}, new Subscription(Prelude["return"](Control_Monad_Eff.monadEff)(Prelude.unit)));
var modifyRVar = function (v) {
    return function (f) {
        return function __do() {
            var _6 = readRVar(v)();
            return writeRVar(v)(f(_6))();
        };
    };
};
var applyComputed = new Prelude.Apply(function (_111) {
    return function (_112) {
        return new Computed({
            read: function __do() {
                var _16 = _111.value0.read();
                var _15 = _112.value0.read();
                return _16(_15);
            }, 
            subscribe: function (ob) {
                return function __do() {
                    var _20 = _111.value0.subscribe(function (f$prime) {
                        return function __do() {
                            var _17 = _112.value0.read();
                            return ob(f$prime(_17))();
                        };
                    })();
                    var _19 = _112.value0.subscribe(function (x$prime) {
                        return function __do() {
                            var _18 = _111.value0.read();
                            return ob(_18(x$prime))();
                        };
                    })();
                    return Prelude["<>"](semigroupSubscription)(_20)(_19);
                };
            }
        });
    };
}, function () {
    return functorComputed;
});
var applicativeComputed = new Prelude.Applicative(function () {
    return applyComputed;
}, function (a) {
    return new Computed({
        read: Prelude.pure(Control_Monad_Eff.applicativeEff)(a), 
        subscribe: function (_) {
            return Prelude.pure(Control_Monad_Eff.applicativeEff)(Data_Monoid.mempty(monoidSubscription));
        }
    });
});
var functorComputed = new Prelude.Functor(Prelude.liftA1(applicativeComputed));
var bindComputed = new Prelude.Bind(function (_109) {
    return function (_110) {
        return new Computed({
            read: function __do() {
                var _7 = _109.value0.read();
                return (function () {
                    var _343 = _110(_7);
                    return _343.value0.read;
                })()();
            }, 
            subscribe: function (ob) {
                return function __do() {
                    var _14 = _109.value0.read();
                    return (function () {
                        var _346 = _110(_14);
                        return function __do() {
                            var _13 = _346.value0.subscribe(ob)();
                            var _12 = Control_Monad_Eff_Ref_Unsafe.unsafeRunRef(Control_Monad_Eff_Ref.newRef(_13))();
                            var _11 = _109.value0.subscribe(function (a$prime) {
                                return function __do() {
                                    var _9 = Control_Monad_Eff_Ref_Unsafe.unsafeRunRef(Control_Monad_Eff_Ref.readRef(_12))();
                                    _9.value0();
                                    var _350 = _110(a$prime);
                                    Prelude[">>="](Control_Monad_Eff.bindEff)(_350.value0.read)(ob)();
                                    var _8 = _350.value0.subscribe(ob)();
                                    return Control_Monad_Eff_Ref_Unsafe.unsafeRunRef(Control_Monad_Eff_Ref.writeRef(_12)(_8))();
                                };
                            })();
                            return Prelude["<>"](semigroupSubscription)(_11)(new Subscription(function __do() {
                                var _10 = Control_Monad_Eff_Ref_Unsafe.unsafeRunRef(Control_Monad_Eff_Ref.readRef(_12))();
                                return _10.value0();
                            }));
                        };
                    })()();
                };
            }
        });
    };
}, function () {
    return applyComputed;
});
var monadComputed = new Prelude.Monad(function () {
    return applicativeComputed;
}, function () {
    return bindComputed;
});
module.exports = {
    Computed: Computed, 
    Inserted: Inserted, 
    Removed: Removed, 
    Subscription: Subscription, 
    Updated: Updated, 
    applicativeComputed: applicativeComputed, 
    applyComputed: applyComputed, 
    bindComputed: bindComputed, 
    functorComputed: functorComputed, 
    insertRArray: insertRArray, 
    modifyRVar: modifyRVar, 
    monadComputed: monadComputed, 
    monoidSubscription: monoidSubscription, 
    newRArray: newRArray, 
    newRVar: newRVar, 
    peekRArray: peekRArray, 
    readComputed: readComputed, 
    readRArray: readRArray, 
    readRVar: readRVar, 
    removeRArray: removeRArray, 
    semigroupSubscription: semigroupSubscription, 
    showArrayChange: showArrayChange, 
    subscribe: subscribe, 
    subscribeArray: subscribeArray, 
    subscribeComputed: subscribeComputed, 
    toComputed: toComputed, 
    toComputedArray: toComputedArray, 
    updateRArray: updateRArray, 
    writeRVar: writeRVar
};
