// Data.Foreign/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Function = require("Data.Function");
var Data_Either = require("Data.Either");
function TypeMismatch(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
TypeMismatch.create = function (value0) {
    return function (value1) {
        return new TypeMismatch(value0, value1);
    };
};
function ErrorAtIndex(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
ErrorAtIndex.create = function (value0) {
    return function (value1) {
        return new ErrorAtIndex(value0, value1);
    };
};
function ErrorAtProperty(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
ErrorAtProperty.create = function (value0) {
    return function (value1) {
        return new ErrorAtProperty(value0, value1);
    };
};
function JSONError(value0) {
    this.value0 = value0;
};
JSONError.create = function (value0) {
    return new JSONError(value0);
};

  function parseJSONImpl(left, right, str) {
    try {
      return right(JSON.parse(str));
    } catch (e) {
      return left(e.toString());
    }
  }
  ;

  function toForeign(value) {
    return value;
  }
  ;

  function unsafeFromForeign(value) {
    return value;
  }
  ;

  function typeOf(value) {
    return typeof value;
  }
  ;

  function tagOf(value) {
    return Object.prototype.toString.call(value).slice(8, -1);
  }
  ;

  function isNull(value) {
    return value === null;
  }
  ;

  function isUndefined(value) {
    return value === undefined;
  }
  ;

  var isArray = Array.isArray || function(value) {
    return Object.prototype.toString.call(value) === '[object Array]';
  };
  ;
var unsafeReadPrim = function (_7) {
    return function (_8) {
        if (tagOf(_8) === _7) {
            return Prelude.pure(Data_Either.applicativeEither)(unsafeFromForeign(_8));
        };
        return new Data_Either.Left(new TypeMismatch(_7, tagOf(_8)));
    };
};
var showForeignError = new Prelude.Show(function (_10) {
    if (_10 instanceof TypeMismatch) {
        return "Type mismatch: expected " + (_10.value0 + (", found " + _10.value1));
    };
    if (_10 instanceof ErrorAtIndex) {
        return "Error at array index " + (Prelude.show(Prelude.showNumber)(_10.value0) + (": " + Prelude.show(showForeignError)(_10.value1)));
    };
    if (_10 instanceof ErrorAtProperty) {
        return "Error at property " + (Prelude.show(Prelude.showString)(_10.value0) + (": " + Prelude.show(showForeignError)(_10.value1)));
    };
    if (_10 instanceof JSONError) {
        return "JSON error: " + _10.value0;
    };
    throw new Error("Failed pattern match");
});
var readString = unsafeReadPrim("String");
var readNumber = unsafeReadPrim("Number");
var readBoolean = unsafeReadPrim("Boolean");
var readArray = function (_9) {
    if (isArray(_9)) {
        return Prelude.pure(Data_Either.applicativeEither)(unsafeFromForeign(_9));
    };
    return new Data_Either.Left(new TypeMismatch("array", tagOf(_9)));
};
var parseJSON = function (json) {
    return parseJSONImpl(Prelude["<<<"](Prelude.semigroupoidArr)(Data_Either.Left.create)(JSONError.create), Data_Either.Right.create, json);
};
var eqForeignError = new Prelude.Eq(function (a) {
    return function (b) {
        return !Prelude["=="](eqForeignError)(a)(b);
    };
}, function (_11) {
    return function (_12) {
        if (_11 instanceof TypeMismatch && _12 instanceof TypeMismatch) {
            return _11.value0 === _12.value0 && _11.value1 === _12.value1;
        };
        if (_11 instanceof ErrorAtIndex && _12 instanceof ErrorAtIndex) {
            return _11.value0 === _12.value0 && Prelude["=="](eqForeignError)(_11.value1)(_12.value1);
        };
        if (_11 instanceof ErrorAtProperty && _12 instanceof ErrorAtProperty) {
            return _11.value0 === _12.value0 && Prelude["=="](eqForeignError)(_11.value1)(_12.value1);
        };
        if (_11 instanceof JSONError && _12 instanceof JSONError) {
            return _11.value0 === _12.value0;
        };
        return false;
    };
});
module.exports = {
    ErrorAtIndex: ErrorAtIndex, 
    ErrorAtProperty: ErrorAtProperty, 
    JSONError: JSONError, 
    TypeMismatch: TypeMismatch, 
    eqForeignError: eqForeignError, 
    isArray: isArray, 
    isNull: isNull, 
    isUndefined: isUndefined, 
    parseJSON: parseJSON, 
    readArray: readArray, 
    readBoolean: readBoolean, 
    readNumber: readNumber, 
    readString: readString, 
    showForeignError: showForeignError, 
    tagOf: tagOf, 
    toForeign: toForeign, 
    typeOf: typeOf, 
    unsafeFromForeign: unsafeFromForeign
};
