// Data.Tuple/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Control_Lazy = require("Control.Lazy");
var Data_Array = require("Data.Array");
var Control_Extend = require("Control.Extend");
var Control_Comonad = require("Control.Comonad");
function Tuple(value0, value1) {
    this.value0 = value0;
    this.value1 = value1;
};
Tuple.create = function (value0) {
    return function (value1) {
        return new Tuple(value0, value1);
    };
};
var zip = Data_Array.zipWith(Tuple.create);
var unzip = function (_29) {
    if (_29.length >= 1) {
        var _181 = _29.slice(1);
        var _175 = unzip(_181);
        return new Tuple(Prelude[":"]((_29[0]).value0)(_175.value0), Prelude[":"]((_29[0]).value1)(_175.value1));
    };
    if (_29.length === 0) {
        return new Tuple([  ], [  ]);
    };
    throw new Error("Failed pattern match");
};
var uncurry = function (_27) {
    return function (_28) {
        return _27(_28.value0)(_28.value1);
    };
};
var swap = function (_30) {
    return new Tuple(_30.value1, _30.value0);
};
var snd = function (_26) {
    return _26.value1;
};
var showTuple = function (__dict_Show_0) {
    return function (__dict_Show_1) {
        return new Prelude.Show(function (_31) {
            return "Tuple (" + (Prelude.show(__dict_Show_0)(_31.value0) + (") (" + (Prelude.show(__dict_Show_1)(_31.value1) + ")")));
        });
    };
};
var semigroupoidTuple = new Prelude.Semigroupoid(function (_36) {
    return function (_37) {
        return new Tuple(_37.value0, _36.value1);
    };
});
var semigroupTuple = function (__dict_Semigroup_2) {
    return function (__dict_Semigroup_3) {
        return new Prelude.Semigroup(function (_38) {
            return function (_39) {
                return new Tuple(Prelude["<>"](__dict_Semigroup_2)(_38.value0)(_39.value0), Prelude["<>"](__dict_Semigroup_3)(_38.value1)(_39.value1));
            };
        });
    };
};
var monoidTuple = function (__dict_Monoid_6) {
    return function (__dict_Monoid_7) {
        return new Data_Monoid.Monoid(function () {
            return semigroupTuple(__dict_Monoid_6["__superclass_Prelude.Semigroup_0"]())(__dict_Monoid_7["__superclass_Prelude.Semigroup_0"]());
        }, new Tuple(Data_Monoid.mempty(__dict_Monoid_6), Data_Monoid.mempty(__dict_Monoid_7)));
    };
};
var functorTuple = new Prelude.Functor(function (_40) {
    return function (_41) {
        return new Tuple(_41.value0, _40(_41.value1));
    };
});
var fst = function (_25) {
    return _25.value0;
};
var lazyLazy1Tuple = function (__dict_Lazy1_9) {
    return function (__dict_Lazy1_10) {
        return new Control_Lazy.Lazy(function (f) {
            return new Tuple(Control_Lazy.defer1(__dict_Lazy1_9)(function (_) {
                return fst(f(Prelude.unit));
            }), Control_Lazy.defer1(__dict_Lazy1_10)(function (_) {
                return snd(f(Prelude.unit));
            }));
        });
    };
};
var lazyLazy2Tuple = function (__dict_Lazy2_11) {
    return function (__dict_Lazy2_12) {
        return new Control_Lazy.Lazy(function (f) {
            return new Tuple(Control_Lazy.defer2(__dict_Lazy2_11)(function (_) {
                return fst(f(Prelude.unit));
            }), Control_Lazy.defer2(__dict_Lazy2_12)(function (_) {
                return snd(f(Prelude.unit));
            }));
        });
    };
};
var lazyTuple = function (__dict_Lazy_13) {
    return function (__dict_Lazy_14) {
        return new Control_Lazy.Lazy(function (f) {
            return new Tuple(Control_Lazy.defer(__dict_Lazy_13)(function (_) {
                return fst(f(Prelude.unit));
            }), Control_Lazy.defer(__dict_Lazy_14)(function (_) {
                return snd(f(Prelude.unit));
            }));
        });
    };
};
var extendTuple = new Control_Extend.Extend(function (_46) {
    return function (_47) {
        return new Tuple(_47.value0, _46(_47));
    };
}, function () {
    return functorTuple;
});
var eqTuple = function (__dict_Eq_15) {
    return function (__dict_Eq_16) {
        return new Prelude.Eq(function (t1) {
            return function (t2) {
                return !Prelude["=="](eqTuple(__dict_Eq_15)(__dict_Eq_16))(t1)(t2);
            };
        }, function (_32) {
            return function (_33) {
                return Prelude["=="](__dict_Eq_15)(_32.value0)(_33.value0) && Prelude["=="](__dict_Eq_16)(_32.value1)(_33.value1);
            };
        });
    };
};
var ordTuple = function (__dict_Ord_4) {
    return function (__dict_Ord_5) {
        return new Prelude.Ord(function () {
            return eqTuple(__dict_Ord_4["__superclass_Prelude.Eq_0"]())(__dict_Ord_5["__superclass_Prelude.Eq_0"]());
        }, function (_34) {
            return function (_35) {
                var _226 = Prelude.compare(__dict_Ord_4)(_34.value0)(_35.value0);
                if (_226 instanceof Prelude.EQ) {
                    return Prelude.compare(__dict_Ord_5)(_34.value1)(_35.value1);
                };
                return _226;
            };
        });
    };
};
var curry = function (f) {
    return function (a) {
        return function (b) {
            return f(new Tuple(a, b));
        };
    };
};
var comonadTuple = new Control_Comonad.Comonad(function () {
    return extendTuple;
}, snd);
var applyTuple = function (__dict_Semigroup_18) {
    return new Prelude.Apply(function (_42) {
        return function (_43) {
            return new Tuple(Prelude["<>"](__dict_Semigroup_18)(_42.value0)(_43.value0), _42.value1(_43.value1));
        };
    }, function () {
        return functorTuple;
    });
};
var bindTuple = function (__dict_Semigroup_17) {
    return new Prelude.Bind(function (_44) {
        return function (_45) {
            var _239 = _45(_44.value1);
            return new Tuple(Prelude["<>"](__dict_Semigroup_17)(_44.value0)(_239.value0), _239.value1);
        };
    }, function () {
        return applyTuple(__dict_Semigroup_17);
    });
};
var applicativeTuple = function (__dict_Monoid_19) {
    return new Prelude.Applicative(function () {
        return applyTuple(__dict_Monoid_19["__superclass_Prelude.Semigroup_0"]());
    }, Tuple.create(Data_Monoid.mempty(__dict_Monoid_19)));
};
var monadTuple = function (__dict_Monoid_8) {
    return new Prelude.Monad(function () {
        return applicativeTuple(__dict_Monoid_8);
    }, function () {
        return bindTuple(__dict_Monoid_8["__superclass_Prelude.Semigroup_0"]());
    });
};
module.exports = {
    Tuple: Tuple, 
    applicativeTuple: applicativeTuple, 
    applyTuple: applyTuple, 
    bindTuple: bindTuple, 
    comonadTuple: comonadTuple, 
    curry: curry, 
    eqTuple: eqTuple, 
    extendTuple: extendTuple, 
    fst: fst, 
    functorTuple: functorTuple, 
    lazyLazy1Tuple: lazyLazy1Tuple, 
    lazyLazy2Tuple: lazyLazy2Tuple, 
    lazyTuple: lazyTuple, 
    monadTuple: monadTuple, 
    monoidTuple: monoidTuple, 
    ordTuple: ordTuple, 
    semigroupTuple: semigroupTuple, 
    semigroupoidTuple: semigroupoidTuple, 
    showTuple: showTuple, 
    snd: snd, 
    swap: swap, 
    uncurry: uncurry, 
    unzip: unzip, 
    zip: zip
};
