// Data.Monoid.First/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var First = {
    create: function (value) {
        return value;
    }
};
var showFirst = function (__dict_Show_0) {
    return new Prelude.Show(function (_140) {
        return "First (" + (Prelude.show(Data_Maybe.showMaybe(__dict_Show_0))(_140) + ")");
    });
};
var semigroupFirst = new Prelude.Semigroup(function (_141) {
    return function (_142) {
        if (_141 instanceof Data_Maybe.Just) {
            return _141;
        };
        return _142;
    };
});
var runFirst = function (_133) {
    return _133;
};
var monoidFirst = new Data_Monoid.Monoid(function () {
    return semigroupFirst;
}, Data_Maybe.Nothing.value);
var eqFirst = function (__dict_Eq_2) {
    return new Prelude.Eq(function (_136) {
        return function (_137) {
            return Prelude["/="](Data_Maybe.eqMaybe(__dict_Eq_2))(_136)(_137);
        };
    }, function (_134) {
        return function (_135) {
            return Prelude["=="](Data_Maybe.eqMaybe(__dict_Eq_2))(_134)(_135);
        };
    });
};
var ordFirst = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqFirst(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_138) {
        return function (_139) {
            return Prelude.compare(Data_Maybe.ordMaybe(__dict_Ord_1))(_138)(_139);
        };
    });
};
module.exports = {
    First: First, 
    eqFirst: eqFirst, 
    monoidFirst: monoidFirst, 
    ordFirst: ordFirst, 
    runFirst: runFirst, 
    semigroupFirst: semigroupFirst, 
    showFirst: showFirst
};
