// Data.Monoid.First/index.js
// Generated by psc-make version 0.6.2
"use strict";
var Prelude = require("Prelude");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var First = {
    create: function (value) {
        return value;
    }
};
var showFirst = function (__dict_Show_0) {
    return new Prelude.Show(function (_222) {
        return "First (" + (Prelude.show(Data_Maybe.showMaybe(__dict_Show_0))(_222) + ")");
    });
};
var semigroupFirst = new Prelude.Semigroup(function (_223) {
    return function (_224) {
        if (_223 instanceof Data_Maybe.Just) {
            return _223;
        };
        return _224;
    };
});
var runFirst = function (_215) {
    return _215;
};
var monoidFirst = new Data_Monoid.Monoid(function () {
    return semigroupFirst;
}, Data_Maybe.Nothing.value);
var eqFirst = function (__dict_Eq_2) {
    return new Prelude.Eq(function (_218) {
        return function (_219) {
            return Prelude["/="](Data_Maybe.eqMaybe(__dict_Eq_2))(_218)(_219);
        };
    }, function (_216) {
        return function (_217) {
            return Prelude["=="](Data_Maybe.eqMaybe(__dict_Eq_2))(_216)(_217);
        };
    });
};
var ordFirst = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqFirst(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_220) {
        return function (_221) {
            return Prelude.compare(Data_Maybe.ordMaybe(__dict_Ord_1))(_220)(_221);
        };
    });
};
module.exports = {
    First: First, 
    eqFirst: eqFirst, 
    monoidFirst: monoidFirst, 
    ordFirst: ordFirst, 
    runFirst: runFirst, 
    semigroupFirst: semigroupFirst, 
    showFirst: showFirst
};
